client-src/elements/utils.ts

```diff
@@ -586,61 +576,75 @@ interface FeatureUpdateInfo {
 
 // Prepare feature/stage changes to be submitted.
 export function formatFeatureChanges(
-  fieldValues,
-  featureId,
+  fieldValues: FieldInfo[],
+  featureId: number,
   formStageId?: number
 ): UpdateSubmitBody {
   let hasChanges = false;
-  const featureChanges = {id: featureId};
-  // Multiple stages can be mutated, so this object is a stage of stages.
-  const stages = {};
-  // When editing an individual stage, always provide stage ID so that
-  // active_stage_id can be set by the server.
-  if (formStageId) {
-    stages[formStageId] = {id: formStageId};
-  }
-  for (const {name, touched, value, stageId, implicitValue} of fieldValues) {
-    // Only submit changes for touched fields or accuracy verification updates.
+  const featureChanges: FeatureUpdateInfo = {id: featureId};
+
+  const directFeatureChanges: { name: string; value: any }[] = [];
+  const pendingStageChanges: {
+    stageId: number;
+    originalName: string;
+    value: any;
+  }[] = [];
+
+  for (const field of fieldValues) {
+    const {name, touched, value, stageId, implicitValue} = field;
     if (!touched) {
       continue;
     }
+    hasChanges = true;
 
-    // Arrays should be submitted as comma-separated strings.
-    let formattedValue = value;
-    if (Array.isArray(formattedValue)) {
-      formattedValue = formattedValue.join(',');
-    }
+    const formattedValue = Array.isArray(value) ? value.join(',') : value;
 
-    // If an explicit value is present, the field value should be truthy.
-    // Otherwise, we ignore the change.
-    // For example, if this is a checkbox to set the active stage, it would need
-    // to be set to true (value), then the active stage would be set to a stage ID (implicitValue).
     if (implicitValue !== undefined) {
-      // Falsey value with an implicit value should be ignored (like an unchecked checkbox).
-      if (!formattedValue) {
-        continue;
+      if (formattedValue) {
+        directFeatureChanges.push({name, value: implicitValue});
       }
-      // fields with implicit values are always changes to feature entities.
-      featureChanges[name] = implicitValue;
     } else if (!stageId) {
-      // If the field doesn't specify a stage ID, that means this change is for a feature field.
-      featureChanges[name] = formattedValue;
+      directFeatureChanges.push({name, value: formattedValue});
     } else {
-      if (!(stageId in stages)) {
-        stages[stageId] = {id: stageId};
-      }
-      stages[stageId][STAGE_FIELD_NAME_MAPPING[name] || name] = {
-        form_field_name: name,
+      pendingStageChanges.push({
+        stageId,
+        originalName: name,
         value: formattedValue,
-      };
+      });
+    }
+  }
+
+  for (const change of directFeatureChanges) {
```

Comment:
I'm not sure why this for-loop is needed.  The existing code was kind of messy, but this proposed change adds a lot of new lines that don't make it simpler.